queue/dispatcher_test.go

```diff
@@ -0,0 +1,294 @@
+package queue
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/mongodb/grip/recovery"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// mockDispatcher provides a mock implementation of a Dispatcher whose behavior
+// can be more easily configured for testing.
+type mockDispatcher struct {
+	queue              amboy.Queue
+	mu                 sync.Mutex
+	dispatched         map[string]dispatcherInfo
+	shouldFailDispatch func(amboy.Job) bool
+	initialLock        func(amboy.Job) error
+	lockPing           func(context.Context, amboy.Job)
+}
+
+func newMockDispatcher(q amboy.Queue) *mockDispatcher {
+	return &mockDispatcher{
+		queue:      q,
+		dispatched: map[string]dispatcherInfo{},
+	}
+}
+
+func (d *mockDispatcher) Dispatch(ctx context.Context, j amboy.Job) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if d.shouldFailDispatch != nil && d.shouldFailDispatch(j) {
+		return errors.Errorf("fail dispatch for job '%s'", j.ID())
+	}
+
+	if _, ok := d.dispatched[j.ID()]; ok {
+		return errors.Errorf("cannot dispatch a job more than once")
+	}
+
+	ti := amboy.JobTimeInfo{
+		Start: time.Now(),
+	}
+	j.UpdateTimeInfo(ti)
+
+	if d.initialLock != nil {
```

Comment:
What is `initialLock` and should we still call `Save` here?

Reply:
`initialLock` is just a mock of taking the distributed locking on the job (i.e. `Lock()` then `Save()`).

Reply:
Ah, so the `initialLock` function would be expected to call `Save()`. Thanks!